Update test library for VS 2019 (#16371)
dotnet/docs: docs/core/tutorials/testing-library-with-visual-studio.md

---
title: Test a .NET Standard class library with .NET Core in Visual Studio
description: Create a unit test project for your .NET Core class library. Verify that your .NET Core class library works correctly with unit tests.
ms.date: 12/23/2019
dev_langs:
  - "csharp"
  - "vb"
ms.custom: "vs-dotnet, seodoc18"
---
# Test a .NET Standard library with .NET Core in Visual Studio

In [Build a .NET Standard library in Visual Studio](library-with-visual-studio.md), you created a simple class library that adds an extension method to the <xref:System.String> class. Now, you'll create a unit test to make sure that it works as expected. You'll add your unit test project to the solution you created in the previous article.

## Create a unit test project

To create the unit test project, do the following:

1. Open the `ClassLibraryProjects` solution you created in the [Build a .NET Standard library in Visual Studio](library-with-visual-studio.md) article.

1. Add a new unit test project named "StringLibraryTest" to the solution.

   1. Right-click on the solution in **Solution Explorer** and select **Add** > **New project**.

>    1. On the **Add a new project** page, enter **mstest** in the search box. Choose **C#** or **Visual Basic** from the Language list, and then choose **All platforms** from the Platform list. Choose the **MsTest Test Project (.NET Core)** template, and then choose **Next**.

Reviewer: We'd perhaps also want to include F# as well? In which case we'd want to change the samples below. Maybe in another PR. @kendrahavens, thoughts? I could put one out after this goes in.

Reviewer 2: Yes, we should list F#. I believe in the past, we did not have F# test templates so we didn't list it.
Since there is no VB sample, I don't think we need to add a F# sample. I don't want to overwhelm the page.

>    1. On the **Configure your new project** page, enter **StringLibraryTest** in the **Project name** box. Then, choose **Create**.

Reviewer 3: ```suggestion
   1. On the **Configure your new project** page, enter **StringLibraryTest** in the **Project name** box. Then choose **Create**.
```

   > [!NOTE]

>    > In addition to an MSTest, you can also create xUnit and nUnit test projects for .NET Core in Visual Studio.

Reviewer 3: This note suggests you might want to add multiple test projects using different frameworks. It would be more helpful if it gave some clues as to why you might want to do that.  That goes beyond the scope of this PR, though.

Author: Very good point and I couldn't find this info in any of our docs here for .NET or VS. Something for us to consider later.

>    - It applies the <xref:Microsoft.VisualStudio.TestTools.UnitTesting.TestMethodAttribute> attribute to define `TestMethod1` as a test method for automatic execution when the unit test is run.

Reviewer 3: "...Attribute attribute" sounds awkward -- this could be changed to "...the [TestMethod] attribute..."

1. In **Solution Explorer**, right-click the **Dependencies** node of the **StringLibraryTest** project and select **Add Reference** from the context menu.

   ![Context menu of StringLibraryTest Dependencies](./media/testing-library-with-visual-studio/add-reference-context-menu.png)

1. In the **Reference Manager** dialog, expand the **Projects** node and check the box next to **StringLibrary**. Adding a reference to the `StringLibrary` assembly allows the compiler to find **StringLibrary** methods. Select the **OK** button. This adds a reference to your class library project, `StringLibrary`.

   ![Visual Studio add project reference dialog](./media/testing-library-with-visual-studio/project-reference-manager.png)

## Add and run unit test methods

When Visual Studio runs a unit test, it executes each method marked with the <xref:Microsoft.VisualStudio.TestTools.UnitTesting.TestMethodAttribute> attribute in a unit test class, the class to which the  <xref:Microsoft.VisualStudio.TestTools.UnitTesting.TestClassAttribute> attribute is applied. A test method ends when the first failure is encountered or when all tests contained in the method have succeeded.

> 

Author: You probably know better than I when it comes to this. I'll change.

Also, do you know where the code for the MSTest SDK comes from? I'm seeing some problems caused by the wrong xml tag being used. I imagine this SDK is auto-generated?!?

> 

Reviewer 4: I think the Assert.ThrowsException is preferred because you can control exactly where in the method you expect the exception to be thrown. With the attribute, the exception could be thrown in an unexpected place and the test would still pass.

For the MSTest SDK code, I'd ask @AbhitejJohn.

> 

Reviewer 4: Suggestion for line 98:

Define three methods, each of which calls an <xref:Microsoft.VisualStudio.TestTools.UnitTesting.Assert> class method repeatedly...

To create the test methods:

1. In the *UnitTest1.cs* or *UnitTest1.vb* code window, replace the code with the following code:

   [!code-csharp[Test#1](~/samples/snippets/csharp/getting_started/with_visual_studio_2017/testlib1.cs)]

>    [!code-vb[Test#1](~/samples/snippets/core/tutorials/vb-library-with-visual-studio/testlib.vb)]

Reviewer: Anyone know where these links point to? They appear to be broken today.

>    Note that your test of uppercase characters in the `TestStartsWithUpper` method includes the Greek capital letter alpha (U+0391) and the Cyrillic capital letter EM (U+041C), and the test of lowercase characters in the `TestDoesNotStartWithUpper` method includes the Greek small letter alpha (U+03B1) and the Cyrillic small letter Ghe (U+0433).

Reviewer 4: ```suggestion
   The test of uppercase characters in the `TestStartsWithUpper` method includes the Greek capital letter alpha (U+0391) and the Cyrillic capital letter EM (U+041C). The test of lowercase characters in the `TestDoesNotStartWithUpper` method includes the Greek small letter alpha (U+03B1) and the Cyrillic small letter Ghe (U+0433).
```

Reviewer 3: Also I would replace the letter names with the actual characters, which can be copied from the code.  Cyrillic "М" is not commonly referred to as "EM" and "г" is not referred to as "Ghe".

Author: Would people understand what we're trying to convey in that sentence if I simply use the actual characters? I'll keep the original for now but happy to change if you think that's better.

```suggestion
   The test of uppercase characters in the `TestStartsWithUpper` method includes Α and М. The test of lowercase characters in the `TestDoesNotStartWithUpper` method includes α and г.
```

   ![Test Explorer window with failing tests](./media/testing-library-with-visual-studio/failed-test-window.png)

> 1. Select the failed test, `TestDoesNotStartWith`. The **Test Explorer** window displays the message produced by the assert: "Assert.IsFalse failed. Expected for 'Error': false; actual: True". Because of the failure, all strings in the array after "Error" were not tested.

Reviewer: nit: Select the failed test, `TestDoesNotStartWith`, under the **Failed Tests** section - to make it more explicit that there is a separate grouping for failed tests.
